src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,399 @@
+                return hkdfExtract(inputKeyMaterial,
+                                   (salt == null) ? null : salt.getEncoded());
+            } catch (InvalidKeyException ike) {
+                throw (InvalidAlgorithmParameterException) new InvalidAlgorithmParameterException(
+                    "an HKDF Extract could not be initialized with the given "
+                    + "key or salt material").initCause(ike);
+            } catch (NoSuchAlgorithmException nsae) {
+                // This is bubbling up from the getInstance of the Mac/Hmac.
+                // Since we're defining these values internally, it is unlikely.
+                throw new ProviderException(
+                    "could not instantiate a Mac with the provided algorithm",
+                    nsae);
+            }
+        } else if (derivationParameterSpec instanceof HKDFParameterSpec.Expand) {
+            HKDFParameterSpec.Expand anExpand =
+                (HKDFParameterSpec.Expand) derivationParameterSpec;
+            // set this value in the "if"
+            if ((pseudoRandomKey = anExpand.prk()) == null) {
+                throw new InvalidAlgorithmParameterException(
+                    "PRK is required for HKDFParameterSpec.Expand");
+            }
+            // set this value in the "if"
+            if ((info = anExpand.info()) == null) {
+                info = new byte[0];
+            }
+            length = anExpand.length();
+            if (length > (hmacLen * 255)) {
+                throw new InvalidAlgorithmParameterException(
+                    "Requested length exceeds maximum allowed key stream "
+                    + "length");
+            }
+            // perform expand
+            try {
+                return Arrays.copyOf(hkdfExpand(pseudoRandomKey, info, length),
+                                     length);
+            } catch (InvalidKeyException ike) {
+                throw (InvalidAlgorithmParameterException) new InvalidAlgorithmParameterException(
+                    "an HKDF Expand could not be initialized with the given "
+                    + "key material").initCause(ike);
+            } catch (NoSuchAlgorithmException nsae) {
+                // This is bubbling up from the getInstance of the Mac/Hmac.
+                // Since we're defining these values internally, it is unlikely.
+                throw new ProviderException(
+                    "could not instantiate a Mac with the provided algorithm",
+                    nsae);
+            }
+        } else if (derivationParameterSpec instanceof HKDFParameterSpec.ExtractThenExpand) {
+            HKDFParameterSpec.ExtractThenExpand anExtractThenExpand =
+                (HKDFParameterSpec.ExtractThenExpand) derivationParameterSpec;
+            ikms = anExtractThenExpand.ikms();
+            salts = anExtractThenExpand.salts();
+            // we should be able to combine these Lists of keys into single
+            // SecretKey Objects,
```

Comment:
"Single SecretKey objects" => "a byte[]"
"List of keys" is really "a list of key segments" which are combined into one key. Same goes for salts. The API is designed with the protocol usage in mind, but the naming we have here does not directly line up with RFC5869 which only mentions singular "IKM" and "SALT".

Reply:
Still has typo "Objects", should be "object".
BTW, I still strongly feel that it can be combined into a byte[] instead of a SecretKey object.